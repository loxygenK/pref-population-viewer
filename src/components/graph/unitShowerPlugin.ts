import Chart, { ChartArea, Plugin as ChartJSPlugin } from "chart.js/auto";

export type UnitAxis = "x" | "y";

export class UnitShowerPlugin implements ChartJSPlugin {
  id = "unitShower";

  readonly axis: UnitAxis;
  readonly content: string;

  constructor(axis: UnitAxis, content: string) {
    this.axis = axis;
    this.content = content;
  }

  beforeRender(chart: Chart) {
    const currentPadding = chart.options.layout?.padding ?? {};

    if (typeof currentPadding !== "object") {
      throw new Error(
        "Padding configuration should be undefined or object if this plugin is used"
      );
    }

    let newPadding: Partial<ChartArea> = {};
    if (this.axis === "x") {
      newPadding = {
        ...currentPadding,
        right: Math.max(30, currentPadding.right ?? 0),
      };
    } else {
      newPadding = {
        ...currentPadding,
        top: Math.max(30, currentPadding.top ?? 0),
      };
    }

    chart.options.layout = {
      ...(chart.options.layout ?? {}),
      padding: newPadding,
    };
  }

  afterDraw(chart: Chart) {
    const canvas = chart.ctx;
    canvas.save();

    const textSize = canvas.measureText(this.content);
    const textWidth = textSize.width;
    const textHeight =
      textSize.actualBoundingBoxDescent + textSize.actualBoundingBoxAscent;

    const x = this.axis === "x" ? chart.chartArea.right + 5 : textWidth;
    const y =
      this.axis === "y"
        ? chart.chartArea.top - textHeight
        : chart.chartArea.bottom + 20;

    canvas.fillText(this.content, x, y);
    canvas.restore();
  }
}
