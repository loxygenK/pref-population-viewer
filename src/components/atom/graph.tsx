import "chart.js/auto";
import { ChartOptions } from "chart.js/auto";
import { Chart } from "react-chartjs-2";
import { generateColor } from "~/util/generateColor";
import { UnitShowerPlugin } from "./graph/unitShowerPlugin";

import React from "react";

export interface GraphData {
  x: number;
  y: number;
}
export interface GraphDataSeries {
  name: string;
  data: Array<GraphData>;
}
export interface GraphProps {
  dataSeries: Array<GraphDataSeries>;
  xUnit: string;
  yUnit: string;
}
export const Graph: React.FC<GraphProps> = ({ dataSeries, xUnit, yUnit }) => {
  const config = React.useMemo(
    () => buildGraphConfiguration(dataSeries),
    [dataSeries]
  );
  return (
    <Chart
      type="line"
      data={config.data}
      options={config.options}
      plugins={[
        new UnitShowerPlugin("x", xUnit),
        new UnitShowerPlugin("y", yUnit),
      ]}
    />
  );
};

const buildGraphConfiguration = (dataSeries: Array<GraphDataSeries>) => {
  if (dataSeries.length === 0) {
    throw new Error("At lease one series must be provided.");
  }

  const data = {
    labels: dataSeries[0].data.map((d) => d.x),
    datasets: dataSeries.map((s) => ({
      label: s.name,
      backgroundColor: generateColor(s.dataIndex),
      borderColor: generateColor(s.dataIndex),
      data: s.data.map((d) => d.y),
    })),
  };

  const options: ChartOptions<"line"> = {
    maintainAspectRatio: false,
    plugins: {
      legend: {
        align: "start",
        position: "right",
        labels: {
          usePointStyle: true,
          pointStyle: "line",
        },
      },
    },
    scales: {
      y: {
        beginAtZero: true,
      },
    },
  };

  return { data, options };
};
