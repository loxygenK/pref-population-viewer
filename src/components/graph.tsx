import "chart.js/auto";
import { ChartData, ChartOptions } from "chart.js/auto";
import { Chart } from "react-chartjs-2";
import { generateColor } from "~/util/generateColor";
import { UnitShowerPlugin } from "./graph/unitShowerPlugin";

export const Graph: React.FC = () => {
  const data: ChartData<"line"> = {
    labels: ["1", "2", "3"],
    datasets: Array.from({ length: 47 }).map((_, i) => ({
      label: `Data set ${i}`,
      backgroundColor: generateColor(i),
      borderColor: generateColor(i),
      data: [10000, 11000 + i * 20, 12000 + i * 20],
    })),
  };

  const options: ChartOptions<"line"> = {
    layout: {
      padding: 30,
    },
    plugins: {
      legend: {
        align: "start",
        position: "right",
        labels: {
          usePointStyle: true,
          pointStyle: "line",
        },
      },
    },
    scales: {
      y: {
        beginAtZero: true,
      },
    },
  };

  return (
    <Chart
      type="line"
      data={data}
      options={options}
      plugins={[
        new UnitShowerPlugin("x", "年度"),
        new UnitShowerPlugin("y", "人口数"),
      ]}
    />
  );
};
