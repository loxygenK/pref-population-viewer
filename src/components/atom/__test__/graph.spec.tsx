import renderer from "react-test-renderer";
import { Graph } from "../graph";

describe("Header component", () => {
  it("renders correcrly", () => {
    const tree = renderer
      .create(
        <Graph
          dataSeries={[
            {
              name: "Hoge",
              data: [
                { x: 2000, y: 100000 },
                { x: 2010, y: 200000 },
                { x: 2020, y: 300000 },
              ],
            },
          ]}
          xUnit="Xunit"
          yUnit="Yunit"
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
