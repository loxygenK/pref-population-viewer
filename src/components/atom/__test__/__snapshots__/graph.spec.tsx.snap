// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders correcrly 1`] = `
<canvas
  height={150}
  role="img"
  width={300}
/>
`;
